alibi_detect/utils/_state.py

```diff
@@ -0,0 +1,71 @@
+"""
+Submodule to handle saving and loading of detector state dictionaries.
+
+TODO: This submodule will eventually be moved to alibi_detect.saving, however this will require legacy save/load
+ support to be refactored or removed, so that the detectors imported in saving._tensorflow.saving/loading do not cause
+ circular dependency issues.
+"""
+import logging
+
+import numpy as np
+from pathlib import Path
+from alibi_detect.utils.pytorch import _save_state_dict as _save_state_dict_pt, \
+    _load_state_dict as _load_state_dict_pt
+
+from alibi_detect.base import BaseDetector
+
+logger = logging.getLogger(__name__)
+
+
+def save_state_dict(detector: BaseDetector, keys: tuple, filepath: Path):
```

Comment:
should this be `StatefulDetectorProtocol`?

Reply:
I struggled to make this change. `mypy` gives:

```
alibi_detect/cd/base_online.py:436: error: Argument 1 to "load_state_dict" has incompatible type "BaseUniDriftOnline"; expected "StatefulDetectorOnline"  [arg-type]
```
`save_state_dict` is called in `BaseUniDriftOnline.save_state`, with the `detector` arg being a `BaseUniDriftOnline` object. For some reason mypy doesn't recognise `BaseUniDriftOnline` as being a `StatefulDetectorOnline` despite it having `save_state` and `load_state` methods...

I'm tempted to leave this as `BaseDetector` since technically `save_state_dict` and `load_state_dict` will actually work for any detector i.e. they just save/load attributes to/from disk, regardless of whether the detector has `save_state` and `load_state` methods... Open to suggestions on how to fix the typing though.

Reply:
urgh, mypy!!!! Yeah lets just leave it for now!

Reply:
🙈